File: typescript/basic/03-translation-service.ts
/**
 * Basic API Example 03: Translation Service
 *
 * A practical example showing how AI agents can provide
 * translation services and get paid automatically.
 *
 * This example demonstrates:
 * - Real-world service pattern
 * - Structured input/output
 * - Budget and timeout configuration
 * - Progress callbacks
 *
 * Run: npx ts-node basic/03-translation-service.ts
 */

import { provide, request } from '@agirails/sdk';
import { waitForProvider } from '../src/utils/helpers';

// Simple mock translation (in production, use actual translation API)
function translate(text: string, from: string, to: string): string {
  const translations: Record<string, Record<string, string>> = {
    'Hello': { de: 'Hallo', es: 'Hola', fr: 'Bonjour', hr: 'Bok' },
    'Thank you': { de: 'Danke', es: 'Gracias', fr: 'Merci', hr: 'Hvala' },
    'Goodbye': { de: 'Auf Wiedersehen', es: 'Adiós', fr: 'Au revoir', hr: 'Doviđenja' },
  };

  return translations[text]?.[to] || `[${to}] ${text}`;
}

async function main() {
  console.log('AGIRAILS Basic API - Translation Service\n');

  // 1. Create translation service provider
  const provider = provide('translation', async (job) => {
    const { text, from, to } = job.input;

    console.log(`  Translating "${text}" from ${from} to ${to}...`);

    // Simulate API call delay
    await new Promise((r) => setTimeout(r, 500));

    const translated = translate(text, from, to);

    return {
      original: text,
      translated,
      from,
      to,
      provider: 'AGIRAILS Demo Translator',
    };
  });

  // Wait for provider to be ready
  await waitForProvider(provider);
  console.log('Translation provider ready:', provider.address);
  console.log('');

  // 2. Request translations
  const phrases = [
    { text: 'Hello', from: 'en', to: 'hr' },
    { text: 'Thank you', from: 'en', to: 'de' },
    { text: 'Goodbye', from: 'en', to: 'es' },
  ];

  for (const phrase of phrases) {
    console.log(`Requesting translation: "${phrase.text}" -> ${phrase.to}`);

    const { result, transaction } = await request('translation', {
      input: phrase,
      budget: 2, // $2 USDC per translation
      timeout: 30000, // 30 seconds timeout
      onProgress: (status) => {
        console.log(`  Status: ${status.state} (${status.progress}%)`);
      },
    });

    const translationResult = result as { original: string; translated: string; from: string; to: string };
    console.log(`  Result: "${translationResult.translated}"`);
    console.log(`  Cost: $${transaction.amount.toFixed(2)} (fee: $${transaction.fee.toFixed(2)})`);
    console.log(`  Duration: ${transaction.duration}ms`);
    console.log('');
  }

  // 3. Show earnings
  console.log('Provider earnings:');
  console.log(`  Total: $${provider.stats.totalEarned.toFixed(2)} USDC`);
  console.log(`  Jobs: ${provider.stats.jobsCompleted}`);

  // 4. Cleanup
  await provider.stop();
  console.log('\nService stopped.');
}

main().catch(console.error);

File: python/basic/03_translation_service.py
#!/usr/bin/env python3
"""
Basic API Example 03: Translation Service

A real-world service example with structured input/output and progress reporting.

This example demonstrates:
- Structured input validation
- Progress callbacks during processing
- Timeout configuration
- Rich result objects

Run: python basic/03_translation_service.py
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.utils.helpers import clear_mock_state, format_usdc


# Simulated translations (in production, use a real API)
TRANSLATIONS = {
    "de": {
        "Hello": "Hallo",
        "World": "Welt",
        "AGIRAILS": "AGIRAILS",
        "is": "ist",
        "awesome": "fantastisch",
    },
    "es": {
        "Hello": "Hola",
        "World": "Mundo",
        "AGIRAILS": "AGIRAILS",
        "is": "es",
        "awesome": "increíble",
    },
    "fr": {
        "Hello": "Bonjour",
        "World": "Monde",
        "AGIRAILS": "AGIRAILS",
        "is": "est",
        "awesome": "génial",
    },
}


async def main() -> None:
    print("AGIRAILS Basic API - Translation Service\n")

    clear_mock_state()

    from agirails import ACTPClient
    from agirails.level0 import (
        provide,
        request,
        set_provider_client,
        start_provider,
        stop_provider,
    )

    REQUESTER = "0x1111111111111111111111111111111111111111"
    PROVIDER = "0x2222222222222222222222222222222222222222"

    # 1. Create translation service
    async def translate_handler(data: dict) -> dict:
        """Translate text to target language."""
        text = data.get("text", "")
        target_lang = data.get("targetLang", "de")
        source_lang = data.get("sourceLang", "en")

        # Get translation dictionary
        translations = TRANSLATIONS.get(target_lang, {})

        # Simulate word-by-word translation
        words = text.split()
        translated_words = []

        for i, word in enumerate(words):
            # Simulate processing time
            await asyncio.sleep(0.1)

            # Translate or keep original
            translated = translations.get(word, word)
            translated_words.append(translated)

        translated_text = " ".join(translated_words)

        return {
            "original": text,
            "translated": translated_text,
            "sourceLang": source_lang,
            "targetLang": target_lang,
            "wordCount": len(words),
        }

    provider = provide(
        "translate",
        translate_handler,
        description="AI-powered translation service",
    )

    # Set up client and provider
    client = await ACTPClient.create(mode="mock", requester_address=REQUESTER)
    await client.mint_tokens(PROVIDER, 10_000_000)
    set_provider_client(client, address=PROVIDER)
    await start_provider()

    print(f"Translation provider started: {PROVIDER[:10]}...")
    print()

    try:
        # 2. Test translations to different languages
        test_cases = [
            {"text": "Hello World", "targetLang": "de"},
            {"text": "AGIRAILS is awesome", "targetLang": "es"},
            {"text": "Hello AGIRAILS", "targetLang": "fr"},
        ]

        for test in test_cases:
            print(f"Translating to {test['targetLang'].upper()}: \"{test['text']}\"")

            # Progress callback
            def on_progress(status):
                if status.progress:
                    print(f"  Progress: {status.progress}%")

            result = await request(
                "translate",
                input=test,
                budget=2.00,  # $2 per translation
                timeout=30000,  # 30 second timeout
                on_progress=on_progress,
                client=client,
                provider=PROVIDER,
            )

            output = result.result
            print(f"  Result: \"{output['translated']}\"")
            print(f"  Words: {output['wordCount']}")
            print(f"  Cost: {format_usdc(result.transaction.amount)}")
            print()

        # 3. Summary
        print("=" * 40)
        print("Translation Service Demo Complete!")
        print(f"Total translations: {len(test_cases)}")
        print("=" * 40)

    finally:
        await stop_provider()

    print("\nDone!")


if __name__ == "__main__":
    asyncio.run(main())
